sklearn/datasets/_base.py

```diff
@@ -1454,28 +1464,160 @@ def _fetch_remote(remote, dirname=None, n_retries=3, delay=1):
 
     Returns
     -------
-    file_path: str
+    file_path: Path
         Full path of the created file.
     """
+    if dirname is None:
+        folder_path = Path(".")
+    else:
+        folder_path = Path(dirname)
+
+    file_path = folder_path / remote.filename
+
+    if file_path.exists():
+        if remote.checksum is None:
+            return file_path
+
+        checksum = _sha256(file_path)
+        if checksum == remote.checksum:
+            return file_path
+        else:
+            warnings.warn(
+                f"SHA256 checksum of existing local file {file_path.name} "
+                f"({checksum}) differs from expected ({remote.checksum}): "
+                f"re-downloading from {remote.url} ."
+            )
+
+    temp_file = NamedTemporaryFile(
+        prefix=remote.filename + ".part_", dir=folder_path, delete=False
+    )
+    temp_file.close()
+    try:
+        temp_file_path = Path(temp_file.name)
+        while True:
+            try:
+                urlretrieve(remote.url, temp_file_path)
+                break
+            except (URLError, TimeoutError):
+                if n_retries == 0:
+                    # If no more retries are left, re-raise the caught exception.
+                    raise
+                warnings.warn(f"Retry downloading from url: {remote.url}")
+                n_retries -= 1
+                time.sleep(delay)
+
+        checksum = _sha256(temp_file_path)
+        if remote.checksum is not None and remote.checksum != checksum:
+            raise OSError(
+                f"The SHA256 checksum of {remote.filename} ({checksum}) "
+                f"differs from expected ({remote.checksum})."
+            )
+    except BaseException:
```

Comment:
Can you add a comment why you are catching `BaseException`? Is this that you want to make sure the temporary file gets deleted in case Control-C gets used during the downoad?

I haven't looked into this in a while, but I seem to remember `except Exception` was advised rather than `except BaseException`